### PR TITLE
Added extract_images_patches impl

### PR DESCRIPTION
**Pull request by SleepProgger**
This PR provides an operation compatible to tensorflows extract_image_patches function. #243
The only case where it provides different results as the tensorflow implementation is when no patch can be extracted, in which case tensorflow produces a shape like [batch, 0, x, depth] and this implementation provides a shape of [0].
It does not honor the data_format of the backend. If that is required that should be easy(ish) to add tho.

Tested with this crude testcase:
```
    BATCHES = 2
    DEPTH = 3
    tf_session = tf.Session()    
    # lol
    for iy in range(1, 6):
        for ix in range(1, 6):
            images = np.random.rand(*(BATCHES, iy, ix, DEPTH))
            k_images = K.variable(images, dtype=images.dtype)
            for ky in range(1, iy):
                for kx in range(1, ix):
                    for sy in range(1, iy):
                        for sx in range(1, ix):
                            for ry in range(1, iy):
                                for rx in range(1, ix):
                                    for padding in ("SAME", "VALID"):
                                        if padding == "VALID" and (ky*ry > iy or kx*rx > ix):
                                            continue
                                        ksizes = [1, ky, kx, 1]
                                        strides = [1, sy, sx, 1]
                                        rates = [1, ry, rx, 1]
                                        print("[%s] img=%s, ksizes=%s, strides=%s, rates=%s, padding=%s" % (
                                            "?",
                                            images.shape,
                                            ksizes, strides,
                                            rates, padding
                                        ), end="\t")
                                        with tf_session.as_default():
                                            tf_patches = tf.extract_image_patches(images, ksizes=ksizes, strides=strides, rates=rates, padding=padding).eval()
                                        k_patches = extract_image_patches(k_images, ksizes, strides, rates, padding).eval()
                                        if np.any(tf_patches != k_patches):
                                            print("\033[91m\033[1mFAILED\033[0m")
                                            print("tf shape: ", tf_patches.shape)
                                            print(tf_patches)
                                            print("k_patches shape: ", k_patches.shape)
                                            print(k_patches)
                                        else:
                                            print("\033[92m\033[1mSUCCESS\033[0m")
```

For the interested here is the same implementation with pure K functions:
```
def extract_image_patches_k(images, ksizes, strides, rates, padding):
    ishape = images.shape.dims
    laxis = ksizes[1]*ksizes[2]*ishape[-1]
    patch_row_eff = ksizes[1] + ((ksizes[1] - 1) * (rates[1] -1))
    patch_col_eff = ksizes[2] + ((ksizes[2] - 1) * (rates[2] -1))
    
    if padding.upper() == "VALID":
        out_rows = math.ceil((ishape[1] - patch_row_eff + 1.) / float(strides[1]))
        out_cols = math.ceil((ishape[2] - patch_col_eff + 1.) / float(strides[2]))
        pad_top = 0
        pad_left = 0
    else:
        out_rows = math.ceil( ishape[1] / float(strides[1]) )
        out_cols = math.ceil( ishape[2] / float(strides[2]) )
        pad_top = max(0, ( (out_rows - 1) * strides[1] + patch_row_eff - ishape[1] ) // 2)
        pad_left = max(0, ( (out_cols - 1) * strides[2] + patch_col_eff - ishape[2] ) // 2)
        # we simply assume padding right + 1 == padding left (same for top/down).
        # This might lead to us padding more as we would need but that won't matter.
        # TF tries to split padding to both sides so left_pad +1 should keep us on the safe side.
        images = K.spatial_2d_padding(images, ((pad_top, pad_top+1), (pad_left, pad_left+1)))

    if out_rows * out_cols == 0:
        return K.constant(np.ndarray(0, dtype=images.shape.dtype.name.lower()))

    if rates[1] != ksizes:
        ksizes[1] = ksizes[1] + (ksizes[1] -1) * (rates[1] - 1)
    if rates[2] != ksizes:
        ksizes[2] = ksizes[2] + (ksizes[2] -1) * (rates[2] - 1)
    ishape = images.shape.dims
    patches = []
    for y in range(out_rows):
        for x in range(out_cols):
            patch = images[:, y*strides[1]:y*strides[1]+ksizes[1]:rates[1], x*strides[2]:x*strides[2]+ksizes[2]:rates[2], :]
            patch = K.reshape(patch, (ishape[0], 1, 1, laxis))
            patches.append(patch)
    patches = K.concatenate(patches, -1)
    nshape = (ishape[0], out_rows, out_cols, laxis)
    patches = K.reshape(patches, nshape)
    return patches
```

### CODE REVIEW

**Comment by SleepProgger** [comment]
Oops, i just realized i forgot to remove `pad_top` and `pad_left` in the padding=="VALID" case.
I'll leave that in as i assume you guys need to merge it manually to your backend version anyway, so you cam simply remove them.

**Comment by brianretford** [comment]
Awesome thank you! BTW we are talking about designing a new version of Tile, would love to include you in those discussions is you're interested

**Comment by SleepProgger** [comment]
Sure, although i have to admit math is not my strongest side.
I also started to love tile to be honest. Needs a bit to get going, but after that point it really is a cool language.

*I am idling in the unofficial plaidml discord in case quicker communication is needed*

**Comment by tzerrell** [comment]
LGTM, I'll point the people working on our CI at this noting it's good to go so it can be merged.

I notice there aren't unit tests -- I checked how it currently works and it seems fine, and for this sort of extension we don't strictly _need_ unit tests so it's ok to merge, but if you want us to notice if we break this you can add some unit tests.

**Comment by SleepProgger** [comment]
Nice to hear that. I am going to add some light unit tests.

**Comment by tzerrell** [comment]
A nice way to add unit tests would be to move this to `op.py` as you suggested and then add a test file for tests of TF-like ops (perhaps `tf_test.py`) to that directory that has a unit test comparing the results of doing this to doing it in pure TF (since this is supposed to reproduce tensorflow behavior).

Adding a test against a fixed expected value to `backend_test.py` would be easier and ok as a basic test (but agreed that it doesn't really match there as you're not comparing the behavior of fixed Keras functions across different backends).

**Comment by SleepProgger** [comment]
Ok, i will move it to op.py and add tf_test.py using a similar structure to backend_test.py.

But i ran into some troubles:
My version doesn't work when width/height are symbolic which is problematic as `y_true` in loss functions is provided as fully symbolic (None all the way).
So i started to move the output shape calculation (NY, NX) into the TILE code, but that did lead to `plaidml.exceptions.Unknown: Variable _T6 used in a context which requires it to be a constant integer` for both the `TMP` sizes and the reshape call.
I assume that makes sense as the size need to be known at bind time.
Is there any way around this ?

A cheap workaround for the loss case is:
```
y_true = K.reshape(y_true, y_pred.shape.dims)
```
so it wouldn't be that bad to have it only work with known sizes, but if there is a way to allow it to work with symbolic sizes i'd highly favorr that.

**Comment by SleepProgger** [comment]
I added unit tests in `tf_test.py` and copied the opTest decorator from `backend_test`.

The `extract_image_patches` function still can't handle symbolic width/height axis and i didn't found a way to do this. If there is a way to do this i'd be happy to change that, otherwise, i'll say this is good enough as is.

**Comment by brianretford** [comment]
integrated in #303
